GoogleForCreators/web-stories-wp
# Floating Menu: Fix Border Color Updates

## Context

Found in bugbash 1.19 - using the eyedropper from the floating menu to update a border color wasn't working. Immediate fix was to remove the eyedropper (#11025) to prevent crashes while getting to the bottom of the inconsistent updates here.

## Summary

This is actually a pretty interesting new case (thanks @littlemilkstudio for spending an hour talking through this and getting to the bottom of it!). Before the floating menus the color picker stays mounted when the eyedropper's in use. Floating menu introduces a new flow where the color picker unmounts. If it's unmounted, the debounced change isn't going to ever get found to actually tell the element to update the color. Since border is the only color picker variant in the floating menu that has the color picker _in_ the popup (as opposed to in the floating menu itself) it's only an issue there. 

## Relevant Technical Choices

`basic` color picker variant: 
- Track if the color picker is mounted with a ref. When it's false and color is changed don't use the debounced change, use the passed in change. The only time the debounce is really needed is on drag interactions. 

`custom` color picker variant: 
- So it's receiving the change handler from the basic color picker, that's why the debounce is actually there - since custom has the dragging on the gradient and slide. That said, it only ever processes that change when a `generatedColor` is true on that existing effect (line 77). This is a problem if the eyedropper's used because (spoiler) it's unmounted now! 
- Move the onChange to a callback of its own, keep existing `updateCurrentColor` to trigger the useColor hook and update as needed and then check if it's unmounted - if it is (and the only way this is true is with floating menu + eyedropper) and there's an rgb key to the value then go ahead and trigger that passed in `handleColorChange` with a manually set color object to propagate changes. 

In the border color floating menu: 
- Fix typo on `hasEyedropper`
- borders can't have gradients 
- borders don't show saved colors (for continuity with sidepanel border color options) 
- added `changedStyle`  for consistency with other border color instance and so that the changedStyle doesn't default to 'background' here. It largely seems to work without it but I figure mirroring the existing border color is good because this whole thing is rather complicated.

## To-do

- Tests probably would be nice, but given the time crunch for release we'll see. 

## User-facing changes

- Can use eyedropper to pick out the border color you want via floating menu. 

https://user-images.githubusercontent.com/10720454/160026202-5df8b424-fbb7-4b80-9cff-7f052bdb6609.mp4



## Testing Instructions

- Please test the color pickers. Make sure that the border color picker in the sidepanel and the floating menu mirror each other. You should be able to set a border color via the eyedropper, color circles, custom gradient, set opacity, type in a hex. 
- Make sure color picker for foreground is fine.
- Make sure color picker from sidepanel also works as expected.




This PR can be tested by following these steps:

1.


## Reviews

### Does this PR have a security-related impact?



### Does this PR change what data or activity we track or use?



### Does this PR have a legal-related impact?



## Checklist




---



Fixes #11024
